mautic/mautic
# Mobile Push Notifications for Native Apps (iOS & Android)

[//]: # ( Please answer the following questions: )

| Q  | A
| --- | ---
| Bug fix? | N
| New feature? | Y
| Related user documentation PR URL | Pending
| Related developer documentation PR URL | Pending
| Issues addressed (#s or URLs) | N/A
| BC breaks? | 
| Deprecations? | 

[//]: # ( Note that all new features should have a related user and/or developer documentation PR in their respective repositories. )

### To Do:

- [x] Update sidebar menu on integration config update
- [x] Additional key/value data fields
- [x] Hide non-essential "advanced" fields

[//]: # ( Required: )
#### Description:

This PR adds the long-awaited feature of pushing notifications to native mobile applications on the iOS and Android platforms (with more platforms to come at a later date). 

As part of this, I've moved the configuration for Twilio & OneSignal from the main configuration page to an integration (this closes #3760)

### iOS Code for OneSignal integration

Prior to using the code examples below, you need to set up the OneSignal SDK in your native app. For iOS, see the [iOS SDK Setup](https://documentation.onesignal.com/docs/ios-sdk-setup) documentation from OneSignal.

To enable push notifications in your iOS app, add the following code (or a variant of it) inside of your `application` func of `AppDelegate`. The code examples below use Swift 3.1. Please modify them to your needs if you're using C#.

```
// Somehow determine the user email. If you have d
let userEmail = "you@domain.com"

OneSignal.initWithLaunchOptions(launchOptions, appId: "YOUR-ONE-SIGNAL-APP-ID")
OneSignal.syncHashedEmail(userEmail);

OneSignal.idsAvailable({(_ userId, _ pushToken) in
    let pushId      = userId != nil ? userId : ""
    let pushEnabled = pushToken != nil ? true : false
    let userData    = UserData(email: userEmail, push_id: pushId!, enabled: pushEnabled)

    self.pushUserDataToMautic(userData, "https://dev.mautic.com/notification/appcallback")
});
```

For ease of use, I've created the following struct and func for sending user data to Mautic.

#### UserData struct
```
struct UserData {
    var email   = String()
    var push_id = String()
    var enabled = Bool()
    
    static func toJSON(_ userData: UserData) -> String {
        let email   = userData.email
        let pushId  = userData.push_id
        let enabled = userData.enabled
            
        return "{\"email\":\"\(email)\",\"push_id\":\"\(pushId)\",\"enabled\":\(enabled)}"
    }
}
```

#### pushUserDataToMautic func

This is a basic function for pushing the UserData struct to your Mautic installation. Modify to meet the needs of your app.

```
func pushUserDataToMautic(_ userData: UserData, _ url: String) {
    var request = URLRequest(url: URL(string: url)!)
    request.httpMethod = "POST"
        
    let postString = UserData.toJSON(userData)
    request.httpBody = postString.data(using: .utf8)
        
    let task = URLSession.shared.dataTask(with: request) { data, response, error in
        guard let data = data, error == nil else {
            // check for fundamental networking error
            return
        }
            
        if let httpStatus = response as? HTTPURLResponse, httpStatus.statusCode != 200 {
            // check for http errors
            return
        }
            
        let responseString = String(data: data, encoding: .utf8)
        let alert = UIAlertController(title: "Response Data", message: responseString, preferredStyle: UIAlertControllerStyle.alert)
        alert.addAction(UIAlertAction(title: "OK", style: UIAlertActionStyle.default, handler: nil))
        self.window?.rootViewController?.present(alert, animated: true, completion: nil);
    }
    task.resume()
}
```

### Notification Stats
In addition to the UserData that gets pushed to Mautic, you can push open / interaction stats to Mautic by sending the UserData struct, with an appended `stat` JSON key

##### Secondary Features

- Menu item access/show based on integration settings
- Integrate `IvoryOrderedFormBundle` to be able to inject/modify entity forms w/listener
- Suppress `no SQL` migration "error" when a migration did not generate any SQL. Used mainly for migrations that simply modified or moved data.

[//]: # ( As applicable: )

#### Steps to test this PR:
1. Have an iOS app with the OneSignal API integrated. To do that, follow the instructions here: [iOS SDK Setup](https://documentation.onesignal.com/docs/ios-sdk-setup)
2. Use the code samples here to set up your native app to push user data to Mautic.
3. Apply this PR to your Mautic and clear cache
4. Add your credentials to the OneSignal Plugin Integration settings page. Be sure to enable Mobile Notifications on the `Feature Settings` tab. Enable iOS or Android based on your app platforms.
5. Open your native app and allow notifications from that app to your device. This will push the UserData struct to Mautic and register the OneSignal player_id to Mautic. This allows Mautic to send push notifications to your device specifically, associated with the correct contact in Mautic (based on the contact email address).
6. Create a new Mobile Notification. Test all functionality (create, edit, delete, pagination, etc)
7. Create a campaign that sends a Mobile Notification to a segment. Add your contact (based on the email provided when you enabled push notifications) to that segment. 
8. Trigger the campaign and see that you've received the push notification to your device.
